Rework main pool alloc to support fragmented buffer. Move zbuffer & fb to different memory banks

aglab2 opened the pull request:
Moves zbuffer and framebuffer to separate memory banks. To implement this, alloc only pool was removed and replaced with main pool instead that supports fragmented memory. Fragmented pool is not using double linked list and is built very similarly to alloc only pool.

Saves 0.5 fps in JRB.
![example](https://github.com/HackerN64/HackerSM64/assets/6104625/7306be24-731c-4688-a2ec-4c8ec45d1c38)

aglab2 committed:
Rework main pool alloc to support fragmented buffer, get rid of alloc only pool, move zbuffer & fb

aglab2 committed:
Move each fb in its own memory bank

aglab2 commented:
Updated PR with splitting FBs to each own memory bank with USE_EXT_MEM. Saved another 0.5f
![image](https://github.com/HackerN64/HackerSM64/assets/6104625/1f158483-cd8a-4b41-96f2-39b7cc3ae6d7)

aglab2 committed:
Added various mode for memory fragmentation, use more conservative fragmentation mode by default

aglab2 committed:
Added another allocation mode that might be beneficial for hacks

arthurtilly commented:
Things that need to be done:

* Make the defines clearer to the average user - ideally slightly less technical and not using magic numbers as each setting. Having fewer settings (only the ones that are necessary) would also be nice.

* Remove `config_memory_private.h`

* Ensure that puppyprint ram usage displays are still entirely accurate

aglab2 commented:
Need to add better documentation for `main_pool_` functions to clearly state their benefits compared to alloc only pool

aglab2 committed:
Remove MAIN_POOL_USE_BEST_FIT to simplify the code, added comments for main pool internal structures

aglab2 committed:
Improved main pool params, delete config_memory_private, fix puppyprint

aglab2 committed:
Use end based main pool region definitions

aglab2 committed:
Implement small alloc optimization for main pool alloc, added single contig region optimization

aglab2 commented:
Finished documentation for main pool allocator. I have also profiled level loads - I have implemented fast path for main_pool_alloc that will be inlined to avoid random slowdowns on level startup.

3.0:
castle inside: 103423 105440 106329 102889 103430
castle grounds: 42575 48902 51084 55313 54764
bob: 71654 70644 70339 70760 71486

3.0 new main pool:
castle inside: 106152 106099 104109 104368 103311
castle grounds: 48520 44482 43405 45028 41257
bob: 70940 70981 70684 70690 71391

I also fixed other issues noted in @arthurtilly comment, code is ready for another round of review

aglab2 committed:
Remove unnecessary options memory layout options, use no fragmentation by default

aglab2 committed:
Document main_pool_alloc better

gheskett commented:
Any chance the z buffer could be moved back to the start of RAM, FB0 to the end of the code segments, and FB1/FB2 bordering 0x80700000? Unless I'm misunderstanding, this would come with the same performance benefits with the least fragmentation necessary.

aglab2 commented:
I used such combination of memory buffers in one of my hacks, I can add it as an option.

arthurtilly commented:
Very much still not happy about this one, especially the overhaul of the memory system. Kaze suggested a way to partially benefit from this optimization without splitting up the main pool by having some buffers at the start of it and some at the end. Worth looking into

gheskett commented:
The layout structure I suggested would only require us to allocate a single buffer at the end of RAM. I think that's gonna be more than enough in the case of allocating dynamic memory, as the rest should be preserved for level data IMO.

gheskett commented:
Replaced by #859
